js/idex.js

```diff
@@ -0,0 +1,969 @@
+                'withdraw': true,
+                'fetchOHLCV': false,
+            },
+            'timeframes': {
+                '1m': 'M1',
+                '3m': 'M3',
+                '5m': 'M5',
+                '15m': 'M15',
+                '30m': 'M30', // default
+                '1h': 'H1',
+                '4h': 'H4',
+                '1d': 'D1',
+                '1w': 'D7',
+                '1M': '1M',
+            },
+            'urls': {
+                'test': 'https://api.idex.market/',
+                'logo': 'https://oracconstimes.com/wp-content/uploads/2018/05/IDEX.market.png',
+                'api': 'https://api.idex.market/',
+                'www': 'https://idex.market/',
+                'doc': [
+                    'https://github.com/AuroraDAO/idex-api-docs',
+                ],
+            },
+            'api': {
+                'public': {
+                    'post': [
+                        'returnTicker',
+                        'returnCurrencies',
+                        'return24Volume',
+                        'returnBalances',
+                        'returnCompleteBalances', // shows amount in orders as well as total
+                        'returnDepositsWithdrawals',
+                        'returnOpenOrders',
+                        'returnOrderBook',
+                        'returnOrderStatus',
+                        'returnOrderTrades',
+                        'returnTradeHistory',
+                        'returnTradeHistoryMeta', // not documented
+                        'returnContractAddress',
+                        'returnNextNonce',
+                    ],
+                },
+                'private': {
+                    'post': [
+                        'order',
+                        'cancel',
+                        'trade',
+                        'withdraw',
+                    ],
+                },
+            },
+            'options': {
+                'contractAddress': undefined,  // 0x2a0c0DBEcC7E4D658f48E01e3fA353F44050c208
+                'orderNonce': undefined,
+            },
+            'exceptions': {
+                '400': BadRequest,
+            },
+            'requiredCredentials': {
+                'walletAddress': true,
+                'privateKey': true,
+                'apiKey': false,
+                'secret': false,
+            },
+        });
+    }
+
+    async fetchMarkets (params = {}) {
+        // idex does not have an endpoint for markets
+        // instead we generate the markets from the endpoint for currencies
+        const markets = await this.publicPostReturnCurrencies ();
+        const result = [];
+        const keys = Object.keys (markets);
+        for (let i = 0; i < keys.length; i++) {
+            const baseId = keys[i];
+            const market = markets[baseId];
+            if (baseId === 'ETH') {
+                continue;
+            }
+            const quote = 'ETH';
```

Comment:
We need to add the following edits:

**Edit 1:**

On the website IDEX supports pairs with the following quote currencies:

- `ETH`
- `TUSD`
- `USDC`
- `EURS`
- `DAI`

And, as you know, ideally, we would not want to hardcode the list ;)

**Edit 2:**

We need to add the contract info for those tokens, namely, `symbol`, `decimals`, and probably all other contract constants as well. Better if we fetch that info, but if it is not available from the exchange itself, we could also take it from etherscan and hardcode it. However, hardcoding really makes a lot of trouble later upon maintenance, when it comes to migrating token contracts and other unpredictable events, like it was, for example, with the TUSD:

- https://etherscan.io/token/0x8dd5fbce2f6a956c3022ba3663759011dd51e73e (old)
- https://etherscan.io/token/0x0000000000085d4780B73119b644AE5ecd22b376 (new)

**Edit 3**:

We have to use `safeCurrencyCode` throughout this file, can't break our own guidelines ) In this case, however, the currency id is used for the currency contract address, so we might have to introduce another field for storing all the props. Maybe we should just move the address to a separate field within the currency structure to keep it all uniform with the rest of the exchanges?